CMakeLists.txt

```diff
@@ -1,16 +1,27 @@
 cmake_minimum_required(VERSION 3.5)
 project(LibreCuda CXX)
 
-# create elf-parser library
-add_subdirectory(third_party/ELFIO)
+set( CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin )
+set( CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib )
 
-add_executable(LibreCuda
-        src/main.cpp
-        src/librecuda.cpp
-        src/librecuda_status.cpp
-        src/cmdqueue.cpp)
 
-target_link_libraries(LibreCuda PRIVATE elfio)
+if(NOT EXISTS ${CMAKE_SOURCE_DIR}/third_party/ELFIO/elfio)
+        message("-- ELFIO not found, fetching ELFIO...")
+        execute_process(COMMAND git submodule update --init)
+else ()
+        message("-- ELFIO found.")
+        add_subdirectory(third_party/ELFIO)
+endif()
 
-# TODO: REMOVE WHEN CRUDE DEBUG MAIN GETS DELETED
-file(COPY ${CMAKE_SOURCE_DIR}/write_float.cubin DESTINATION ${CMAKE_BINARY_DIR})
\ No newline at end of file
+set(
+    DRIVERAPI_SOURCES
+    ${CMAKE_SOURCE_DIR}/src/librecuda.cpp
+    ${CMAKE_SOURCE_DIR}/src/librecuda_status.cpp
+    ${CMAKE_SOURCE_DIR}/src/cmdqueue.cpp
+)
+
+add_library(driverapi SHARED ${DRIVERAPI_SOURCES})
```

Comment:
As stated earlier, this should probably be moved to its own driverapi/CMakeLists.txt file along with src/
include/ and internal/.
include/nvidia also shouldn't be available for inclusion through driverapi and should be considered internal headers.

Reply:
.cpp files should probably be in driverapi/src

Reply:
Also, librecuda.h is the only public header you should be able to include. hence it should be in driverapi/include
nvidia/ is considered internal.
Remember, include/ is PUBLIC
internal/ is PRIVATE

Reply:
OK

Reply:
Now the driverapi/ seems more cleaner(right?)
```text
../driverapi/
├── CMakeLists.txt
├── include
│   ├── cmdqueue.h
│   ├── librecuda_all_statuess.h
│   ├── librecuda.h
│   ├── librecuda_internal.h
│   ├── librecuda_status.h
│   └── librecuda_status_internal.h
├── internal
│   └── nvidia
│       ├── alloc_channel.h
Nvidia.balabala.h
└── src
    ├── cmdqueue.cpp
    ├── librecuda.cpp
    └── librecuda_status.cpp
```
the whole project structure look like this:
```text
 ______________
< cow bless me >
 --------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

┌ |base|💁  per1cycle @ 💻  DESKTOP-LTJGNCK in 📁  build on 🌿  master ✓
└❯ tree ..
..
├── build
│   ├── bin
│   │   ├── test_write_float
│   │   └── write_float.cubin
│   ├── CMakeCache.txt
│   ├── CMakeFiles
│   │   ├── 3.22.1
│   │   │   ├── CMakeCCompiler.cmake
│   │   │   ├── CMakeCXXCompiler.cmake
│   │   │   ├── CMakeDetermineCompilerABI_C.bin
│   │   │   ├── CMakeDetermineCompilerABI_CXX.bin
│   │   │   ├── CMakeSystem.cmake
│   │   │   ├── CompilerIdC
│   │   │   │   ├── a.out
│   │   │   │   ├── CMakeCCompilerId.c
│   │   │   │   └── tmp
│   │   │   └── CompilerIdCXX
│   │   │       ├── a.out
│   │   │       ├── CMakeCXXCompilerId.cpp
│   │   │       └── tmp
│   │   ├── cmake.check_cache
│   │   ├── CMakeDirectoryInformation.cmake
│   │   ├── CMakeOutput.log
│   │   ├── CMakeTmp
│   │   ├── Makefile2
│   │   ├── Makefile.cmake
│   │   ├── progress.marks
│   │   └── TargetDirectories.txt
│   ├── cmake_install.cmake
│   ├── driverapi
│   │   ├── CMakeFiles
│   │   │   ├── CMakeDirectoryInformation.cmake
│   │   │   ├── driverapi.dir
│   │   │   │   ├── build.make
│   │   │   │   ├── cmake_clean.cmake
│   │   │   │   ├── compiler_depend.make
│   │   │   │   ├── compiler_depend.ts
│   │   │   │   ├── DependInfo.cmake
│   │   │   │   ├── depend.make
│   │   │   │   ├── flags.make
│   │   │   │   ├── link.txt
│   │   │   │   ├── progress.make
│   │   │   │   └── src
│   │   │   │       ├── cmdqueue.cpp.o
│   │   │   │       ├── cmdqueue.cpp.o.d
│   │   │   │       ├── librecuda.cpp.o
│   │   │   │       ├── librecuda.cpp.o.d
│   │   │   │       ├── librecuda_status.cpp.o
│   │   │   │       └── librecuda_status.cpp.o.d
│   │   │   └── progress.marks
│   │   ├── cmake_install.cmake
│   │   └── Makefile
│   ├── lib
│   │   └── libdriverapi.so
│   ├── Makefile
│   ├── tests
│   │   ├── CMakeFiles
│   │   │   ├── CMakeDirectoryInformation.cmake
│   │   │   └── progress.marks
│   │   ├── cmake_install.cmake
│   │   ├── Makefile
│   │   └── write_float
│   │       ├── CMakeFiles
│   │       │   ├── CMakeDirectoryInformation.cmake
│   │       │   ├── progress.marks
│   │       │   └── test_write_float.dir
│   │       │       ├── build.make
│   │       │       ├── cmake_clean.cmake
│   │       │       ├── compiler_depend.make
│   │       │       ├── compiler_depend.ts
│   │       │       ├── DependInfo.cmake
│   │       │       ├── depend.make
│   │       │       ├── flags.make
│   │       │       ├── link.txt
│   │       │       ├── main.cpp.o
│   │       │       ├── main.cpp.o.d
│   │       │       └── progress.make
│   │       ├── cmake_install.cmake
│   │       └── Makefile
│   └── third_party
│       └── ELFIO
│           ├── CMakeFiles
│           │   ├── CMakeDirectoryInformation.cmake
│           │   └── progress.marks
│           ├── cmake_install.cmake
│           ├── elfio_version.hpp.copy
│           └── Makefile
├── CMakeLists.txt
├── driverapi
│   ├── CMakeLists.txt
│   ├── include
│   │   ├── cmdqueue.h
│   │   ├── librecuda_all_statuess.h
│   │   ├── librecuda.h
│   │   ├── librecuda_internal.h
│   │   ├── librecuda_status.h
│   │   └── librecuda_status_internal.h
│   ├── internal
│   │   └── nvidia
│   │       ├── alloc_channel.h
│   │       ├── cc_drv.h
│   │       ├── cl0040.h
│   │       ├── cl0080.h
│   │       ├── cl0080_notification.h
│   │       ├── cl2080.h
│   │       ├── cl2080_notification.h
│   │       ├── cl9010_callback.h
│   │       ├── cl9010.h
│   │       ├── cl9067.h
│   │       ├── cl90f1.h
│   │       ├── cla06c.h
│   │       ├── cla0c0qmd.h
│   │       ├── clc3c0qmd.h
│   │       ├── clc461.h
│   │       ├── clc56f.h
│   │       ├── clc5c0qmd.h
│   │       ├── clc6b5.h
│   │       ├── clc6c0.h
│   │       ├── clc6c0qmd.h
│   │       ├── cpuopsys.h
│   │       ├── ctrl0000base.h
│   │       ├── ctrl0000gpu.h
│   │       ├── ctrl0000system.h
│   │       ├── ctrl0073base.h
│   │       ├── ctrl0073system.h
│   │       ├── ctrl0080base.h
│   │       ├── ctrl0080bsp.h
│   │       ├── ctrl0080gpu.h
│   │       ├── ctrl0080gr.h
│   │       ├── ctrl0080msenc.h
│   │       ├── ctrl2080base.h
│   │       ├── ctrl2080fifo.h
│   │       ├── ctrl2080gpu.h
│   │       ├── ctrl2080gr.h
│   │       ├── ctrl2080internal.h
│   │       ├── ctrl30f1.h
│   │       ├── ctrl906f.h
│   │       ├── ctrl90f1.h
│   │       ├── ctrla06c.h
│   │       ├── ctrla06fbase.h
│   │       ├── ctrla06fevent.h
│   │       ├── ctrla06fgpfifo.h
│   │       ├── ctrla06f.h
│   │       ├── ctrla06finternal.h
│   │       ├── ctrlc36f.h
│   │       ├── ctrlxxxx.h
│   │       ├── dev_mmu.h
│   │       ├── g_allclasses.h
│   │       ├── mmu_fmt_types.h
│   │       ├── nvcfg_sdk.h
│   │       ├── nvCpuUuid.h
│   │       ├── nv_escape.h
│   │       ├── nvgputypes.h
│   │       ├── nvimpshared.h
│   │       ├── nv-ioctl.h
│   │       ├── nv-ioctl-numbers.h
│   │       ├── nvlimits.h
│   │       ├── nvmisc.h
│   │       ├── nvos.h
│   │       ├── nvstatuscodes.h
│   │       ├── nvstatus.h
│   │       ├── nvtypes.h
│   │       ├── nv-unix-nvos-params-wrappers.h
│   │       ├── rs_access.h
│   │       ├── uvm_ioctl.h
│   │       ├── uvm_linux_ioctl.h
│   │       └── uvm_types.h
│   └── src
│       ├── cmdqueue.cpp
│       ├── librecuda.cpp
│       └── librecuda_status.cpp
├── LICENSE
├── README.MD
├── tests
│   ├── CMakeLists.txt
│   ├── compile_cubin.sh
│   ├── complex
│   │   ├── CMakeLists.txt
│   │   └── complex.ptx
│   └── write_float
│       ├── CMakeLists.txt
│       ├── main.cpp
│       ├── write_float.asm
│       ├── write_float.cu
│       ├── write_float.cubin
│       └── write_float.ptx
└── third_party
    └── ELFIO
        ├── cmake
        │   └── elfioConfig.cmake.in
        ├── CMakeLists.txt
        ├── doc
        │   ├── elfio.docx
        │   ├── elfio.pdf
        │   ├── images
        │   │   ├── annot-close.png
        │   │   ├── annot-open.png
        │   │   ├── blank.png
        │   │   ├── callouts
        │   │   │   ├── 10.gif
        │   │   │   ├── 10.png
        │   │   │   ├── 10.svg
        │   │   │   ├── 11.gif
        │   │   │   ├── 11.png
        │   │   │   ├── 11.svg
        │   │   │   ├── 12.gif
        │   │   │   ├── 12.png
        │   │   │   ├── 12.svg
        │   │   │   ├── 13.gif
        │   │   │   ├── 13.png
        │   │   │   ├── 13.svg
        │   │   │   ├── 14.gif
        │   │   │   ├── 14.png
        │   │   │   ├── 14.svg
        │   │   │   ├── 15.gif
        │   │   │   ├── 15.png
        │   │   │   ├── 15.svg
        │   │   │   ├── 16.svg
        │   │   │   ├── 17.svg
        │   │   │   ├── 18.svg
        │   │   │   ├── 19.svg
        │   │   │   ├── 1.gif
        │   │   │   ├── 1.png
        │   │   │   ├── 1.svg
        │   │   │   ├── 20.svg
        │   │   │   ├── 21.svg
        │   │   │   ├── 22.svg
        │   │   │   ├── 23.svg
        │   │   │   ├── 24.svg
        │   │   │   ├── 25.svg
        │   │   │   ├── 26.svg
        │   │   │   ├── 27.svg
        │   │   │   ├── 28.svg
        │   │   │   ├── 29.svg
        │   │   │   ├── 2.gif
        │   │   │   ├── 2.png
        │   │   │   ├── 2.svg
        │   │   │   ├── 30.svg
        │   │   │   ├── 3.gif
        │   │   │   ├── 3.png
        │   │   │   ├── 3.svg
        │   │   │   ├── 4.gif
        │   │   │   ├── 4.png
        │   │   │   ├── 4.svg
        │   │   │   ├── 5.gif
        │   │   │   ├── 5.png
        │   │   │   ├── 5.svg
        │   │   │   ├── 6.gif
        │   │   │   ├── 6.png
        │   │   │   ├── 6.svg
        │   │   │   ├── 7.gif
        │   │   │   ├── 7.png
        │   │   │   ├── 7.svg
        │   │   │   ├── 8.gif
        │   │   │   ├── 8.png
        │   │   │   ├── 8.svg
        │   │   │   ├── 9.gif
        │   │   │   ├── 9.png
        │   │   │   └── 9.svg
        │   │   ├── caution.gif
        │   │   ├── caution.png
        │   │   ├── caution.svg
        │   │   ├── caution.tif
        │   │   ├── colorsvg
        │   │   │   ├── caution.svg
        │   │   │   ├── home.svg
        │   │   │   ├── important.svg
        │   │   │   ├── next.svg
        │   │   │   ├── note.svg
        │   │   │   ├── prev.svg
        │   │   │   ├── tip.svg
        │   │   │   ├── up.svg
        │   │   │   └── warning.svg
        │   │   ├── draft.png
        │   │   ├── home.gif
        │   │   ├── home.png
        │   │   ├── home.svg
        │   │   ├── important.gif
        │   │   ├── important.png
        │   │   ├── important.svg
        │   │   ├── important.tif
        │   │   ├── next.gif
        │   │   ├── next.png
        │   │   ├── next.svg
        │   │   ├── note.gif
        │   │   ├── note.png
        │   │   ├── note.svg
        │   │   ├── note.tif
        │   │   ├── prev.gif
        │   │   ├── prev.png
        │   │   ├── prev.svg
        │   │   ├── res
        │   │   │   ├── 20231119165006-100.png
        │   │   │   └── title.png
        │   │   ├── res2
        │   │   │   ├── 20231119095119.jpg
        │   │   │   ├── 20231119095351.jpg
        │   │   │   ├── 20231119095425.jpg
        │   │   │   ├── 20231119095558.jpg
        │   │   │   ├── 20231119095613.jpg
        │   │   │   ├── 20231119095721.jpg
        │   │   │   ├── 20231119095920.jpg
        │   │   │   ├── 20231119095932.jpg
        │   │   │   ├── 20231119095944.jpg
        │   │   │   ├── 20231119100133.jpg
        │   │   │   ├── 20231119100142.jpg
        │   │   │   ├── 20231119165006-1.jpg
        │   │   │   ├── 20231119165006-256.png
        │   │   │   ├── 20231119165006.jpg
        │   │   │   ├── 20231119165016.jpg
        │   │   │   ├── 20231119165024.jpg
        │   │   │   ├── 20231119165842.jpg
        │   │   │   ├── 20231119165851.jpg
        │   │   │   ├── 20231119165907.jpg
        │   │   │   ├── 20231119165912.jpg
        │   │   │   └── iceland.jpeg
        │   │   ├── tip.gif
        │   │   ├── tip.png
        │   │   ├── tip.svg
        │   │   ├── tip.tif
        │   │   ├── toc-blank.png
        │   │   ├── toc-minus.png
        │   │   ├── toc-plus.png
        │   │   ├── up.gif
        │   │   ├── up.png
        │   │   ├── up.svg
        │   │   ├── warning.gif
        │   │   ├── warning.png
        │   │   ├── warning.svg
        │   │   └── warning.tif
        │   └── site
        │       ├── index.htm
        │       └── style.css
        ├── elfio
        │   ├── elfio_array.hpp
        │   ├── elfio_dump.hpp
        │   ├── elfio_dynamic.hpp
        │   ├── elfio_header.hpp
        │   ├── elfio.hpp
        │   ├── elfio_modinfo.hpp
        │   ├── elfio_note.hpp
        │   ├── elfio_relocation.hpp
        │   ├── elfio_section.hpp
        │   ├── elfio_segment.hpp
        │   ├── elfio_strings.hpp
        │   ├── elfio_symbols.hpp
        │   ├── elfio_utils.hpp
        │   ├── elfio_version.hpp
        │   ├── elfio_versym.hpp
        │   └── elf_types.hpp
        ├── examples
        │   ├── add_section
        │   │   ├── add_section.cpp
        │   │   └── CMakeLists.txt
        │   ├── anonymizer
        │   │   ├── anonymizer.cpp
        │   │   └── CMakeLists.txt
        │   ├── CMakeLists.txt
        │   ├── c_wrapper
        │   │   ├── c_example.c
        │   │   ├── CMakeLists.txt
        │   │   ├── elfio_c_wrapper.cpp
        │   │   └── elfio_c_wrapper.h
        │   ├── elfdump
        │   │   ├── CMakeLists.txt
        │   │   └── elfdump.cpp
        │   ├── proc_mem
        │   │   ├── CMakeLists.txt
        │   │   └── proc_mem.cpp
        │   ├── sudo_gdb.sh
        │   ├── tutorial
        │   │   ├── CMakeLists.txt
        │   │   └── tutorial.cpp
        │   ├── write_obj
        │   │   ├── CMakeLists.txt
        │   │   └── write_obj.cpp
        │   └── writer
        │       ├── CMakeLists.txt
        │       └── writer.cpp
        ├── LICENSE.txt
        ├── README.md
        └── tests
            ├── CMakeLists.txt
            ├── elf_examples
            │   ├── 64bitLOAD.elf
            │   ├── ARMSCII-8.so
            │   ├── arm_v7m_test_debug.elf
            │   ├── arm_v7m_test_release.elf
            │   ├── asm
            │   ├── asm64
            │   ├── asm64.lst
            │   ├── asm64.o
            │   ├── asm.lst
            │   ├── asm.o
            │   ├── asm.readelf
            │   ├── asm.s
            │   ├── crash-060833f08dc14d1712428742b3cad7af17b36bb7
            │   ├── crash-7d695153fd8052529d480c2352d4ada33a44bada
            │   ├── crash-b82f05b0b25c8fdc98480e6d76b6d5f9164ae2bc
            │   ├── crash-e1ce7cecf01cf800397a4302854d9d76fa19763c
            │   ├── crash-e3c41070342cf84dea077356ddbb8ebf4326a601
            │   ├── crash.elf
            │   ├── ctors
            │   ├── elf_header_i386_32.elf
            │   ├── entropy.so
            │   ├── hello_32
            │   ├── hello_32.o
            │   ├── hello_32_o.txt
            │   ├── hello_32.txt
            │   ├── hello_64
            │   ├── hello_64.o
            │   ├── hello_64_o.txt
            │   ├── hello_64.txt
            │   ├── hello_arm
            │   ├── hello_arm.o
            │   ├── hello_arm_stripped
            │   ├── hello.c
            │   ├── helloworld.rpx
            │   ├── libfunc32.so
            │   ├── libfunc.so
            │   ├── main
            │   ├── main32
            │   ├── mismatched_segments.elf
            │   ├── null_endianess_convertor
            │   ├── ppc-32bit-specimen3.elf
            │   ├── read_write_arm_elf32_input
            │   ├── startup.eln
            │   ├── startup_orig.eln
            │   ├── test_ppc
            │   ├── test_ppc.cpp
            │   ├── test_ppc.o
            │   ├── test_ppc_o.txt
            │   ├── test_ppc.txt
            │   ├── write_exe_i386_32_match
            │   ├── write_exe_i386_32_work_dump.txt
            │   ├── write_obj_i386_32_match.o
            │   ├── write_obj_i386_64_match.o
            │   ├── x86_64_static
            │   └── zavl.ko
            ├── elfio_fuzzer.cpp
            ├── ELFIOTest1.cpp
            ├── ELFIOTest2.cpp
            ├── ELFIOTest.cpp
            └── runELFtests

52 directories, 392 files
```

Reply:
This looks correct

Reply:
great!